### PR TITLE
AP_Landing: option to keep landing throttle at thr_min during flare

### PR DESCRIPTION
**Pull request by davidbuzz**

### CODE REVIEW

File: ArduPlane/servos.cpp [hunk 1]
```diff
@@ -445,9 +445,7 @@ void Plane::set_servos_controlled(void)
         } else {
             max_throttle = aparm.throttle_max;
         }
-    } else if (landing.is_flaring()) {
-        min_throttle = 0;
-    }
```
Comment (tridge) [inline ArduPlane/servos.cpp:450]: please restore those lines

File: ArduPlane/servos.cpp [hunk 2]
```diff
@@ -469,8 +467,9 @@ void Plane::set_servos_controlled(void)
             SRV_Channels::set_output_scaled(SRV_Channel::k_throttleRight, 0);
         }
     } else if (suppress_throttle()) {
-        // throttle is suppressed in auto mode
-        SRV_Channels::set_output_scaled(SRV_Channel::k_throttle, 0);
+        SRV_Channels::set_output_scaled(SRV_Channel::k_throttle, 0 ); // default
+        // throttle is suppressed (above) to zero in final flare in auto mode, but we allow instead thr_min if user prefers, eg turbines:
+        if (landing.is_flaring() && landing.use_thr_min_during_flare() )  SRV_Channels::set_output_scaled(SRV_Channel::k_throttle, aparm.throttle_min.get());
```
Comment (tridge) [inline ArduPlane/servos.cpp:472]: please follow coding style with { and whitespace

File: libraries/AP_Landing/AP_Landing.cpp [hunk 3]
```diff
@@ -142,6 +142,13 @@ const AP_Param::GroupInfo AP_Landing::var_info[] = {
     // @Group: DS_
     // @Path: AP_Landing_Deepstall.cpp
     AP_SUBGROUPINFO(deepstall, "DS_", 15, AP_Landing, AP_Landing_Deepstall),
+
+    // @Param: _OPTIONS
+    // @DisplayName: Landing options bitmask
+    // @Description: Bitmask of options to use with landing.
+    // @Bitmask: 0: normal landing behaviour 1: honor min_throttle_during_landing , other bits tbd
```
Comment (tridge) [inline libraries/AP_Landing/AP_Landing.cpp:149]: incorrect bitmask docs

File: libraries/AP_Landing/AP_Landing.cpp [hunk 4]
```diff
@@ -606,6 +613,11 @@ bool AP_Landing::is_throttle_suppressed(void) const
     }
 }
 
+//defaults to false, but _options bit zero enables it.
+bool AP_Landing::use_thr_min_during_flare(void){
```
Comment (tridge) [inline libraries/AP_Landing/AP_Landing.cpp:617]: whitespace

File: libraries/AP_Landing/AP_Landing.cpp [hunk 5]
```diff
@@ -606,6 +613,11 @@ bool AP_Landing::is_throttle_suppressed(void) const
     }
 }
 
+//defaults to false, but _options bit zero enables it.
+bool AP_Landing::use_thr_min_during_flare(void){ 
+    return OptionsMask::ON_LANDING_FLARE_USE_THR_MIN & _options;
```
Comment (tridge) [inline libraries/AP_Landing/AP_Landing.cpp:618]: when checking a bit to make a booean, use 
  return (bitmask & XXX) != 0;

File: libraries/AP_Landing/AP_Landing.h [hunk 6]
```diff
@@ -57,6 +57,12 @@ class AP_Landing {
 //      TODO: TYPE_HELICAL,
     };
 
+    // we support upto 32 boolean bits for users wanting to change landing behaviour. 
+    enum OptionsMask {
+        ON_LANDING_FLARE_USE_THR_MIN                   = (1<<0),   // If set then set trottle to thr_min instead of zero on final flare
+        //ON_BLAH_BLAH         = (1<<1),   // If set then
```
Comment (tridge) [inline libraries/AP_Landing/AP_Landing.h:63]: please no comments like this

File: libraries/AP_Landing/AP_Landing.h [hunk 7]
```diff
@@ -72,6 +78,7 @@ class AP_Landing {
     bool is_ground_steering_allowed(void) const;
     bool is_throttle_suppressed(void) const;
     bool is_flying_forward(void) const;
+    bool use_thr_min_during_flare(void); //defaults to false, but _options bit zero enables it.
```
Comment (tridge) [inline libraries/AP_Landing/AP_Landing.h:81]: should be const fn

File: libraries/AP_Landing/AP_Landing.h [hunk 8]
```diff
@@ -115,6 +122,8 @@ class AP_Landing {
         bool in_progress:1;
     } flags;
 
+    AP_Int32 _options;    // user-configurable bitmask options, via a parameter, for landing
```
Comment (tridge) [inline libraries/AP_Landing/AP_Landing.h:125]: I think 16 bits is enough

Comment (davidbuzz) [reply libraries/AP_Landing/AP_Landing.h:124]: fixed in latest push

File: libraries/AP_Landing/AP_Landing.h (hunk 7 above)
Comment (davidbuzz) [reply libraries/AP_Landing/AP_Landing.h:80]: fixed in latest push

File: libraries/AP_Landing/AP_Landing.cpp (hunk 5 above)
Comment (davidbuzz) [reply libraries/AP_Landing/AP_Landing.cpp]: fixed in latest push

File: libraries/AP_Landing/AP_Landing.cpp (hunk 4 above)
Comment (davidbuzz) [reply libraries/AP_Landing/AP_Landing.cpp]: fixed in latest push

File: ArduPlane/servos.cpp (hunk 2 above)
Comment (davidbuzz) [reply ArduPlane/servos.cpp]: fixed in latest push

**Comment by davidbuzz** [comment]
> needs split to 2 commits

done.

**Comment by tridge** [comment]
CI fail:
    Parameters (Test Parameter Set/Get) (Downloaded parameters missing in XML) (see /home/travis/build/ArduPilot/buildlogs/QuadPlane-Parameters.txt)

File: libraries/AP_Landing/AP_Landing.cpp [hunk 9]
```diff
@@ -142,7 +142,14 @@ const AP_Param::GroupInfo AP_Landing::var_info[] = {
     // @Group: DS_
     // @Path: AP_Landing_Deepstall.cpp
     AP_SUBGROUPINFO(deepstall, "DS_", 15, AP_Landing, AP_Landing_Deepstall),
-    
+
+    // @Param: _OPTIONS
```
Comment (tridge) [inline libraries/AP_Landing/AP_Landing.cpp:146]: extra _ in the @Param, which is why CI is failing

Comment (davidbuzz) [reply libraries/AP_Landing/AP_Landing.cpp:146]: ... so close.

**Pull request merged**